File: Diagrammatize.Design.Pattern/code/python/Proxy/A1/PrinterManager.py

class PrinterManager:
    factory_method_map = dict()

    @staticmethod
    def newInstance(className):
        factory_method = PrinterManager.factory_method_map.get(className, None)
        if factory_method is None:
            return None
        return factory_method()

    @staticmethod
    def registerFactoryMethod(className, factory_method):
        PrinterManager.factory_method_map[className] = factory_method

File: Diagrammatize.Design.Patterns/code/cxx/Proxy/A1/PrinterManager.cpp
#include "PrinterManager.hpp"

PrinterManager::FactoryMethodMapType& PrinterManager::getFactoryMethodMap() {
    static FactoryMethodMapType factory_method_map{};
    return factory_method_map;
}

void PrinterManager::registerFactoryMethod(const std::string& className, FactoryMethodType factory_method) {
    auto& factory_method_map = getFactoryMethodMap();
    factory_method_map[className] = factory_method;
}

Printable* PrinterManager::newInstance(const std::string& className) {
    auto& factory_method_map = getFactoryMethodMap();
    auto iter = factory_method_map.find(className);
    if (iter == factory_method_map.end()) {
        return nullptr;
    }
    auto factory_method = iter->second;
    return factory_method();
}

PrinterManager::RegisterFactoryMethodHelper::RegisterFactoryMethodHelper(
        const std::string& className, PrinterManager::FactoryMethodType factory_method) {
    PrinterManager::registerFactoryMethod(className, factory_method);
}
